File: SnakeGameCSharp/Food.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace SnakeGameCSharp
{
    public class Food
    {
        public (int X, int Y) Position { get; private set; }
        private readonly int _boardWidth;
        private readonly int _boardHeight;
        private readonly Random _random = new Random();

        public Food(int boardWidth, int boardHeight, List<(int X, int Y)> snakeBody)
        {
            _boardWidth = boardWidth;
            _boardHeight = boardHeight;
            GenerateNewPosition(snakeBody);
        }

        public void GenerateNewPosition(List<(int X, int Y)> snakeBody)
        {
            while (true)
            {
                int x = _random.Next(0, _boardWidth);
                int y = _random.Next(0, _boardHeight);

                // Check if the new position is on the snake
                bool onSnake = false;
                foreach (var segment in snakeBody)
                {
                    if (segment.X == x && segment.Y == y)
                    {
                        onSnake = true;
                        break;
                    }
                }

                if (!onSnake)
                {
                    Position = (x, y);
                    break;
                }
            }
        }
    }
}

File: snake-game-react-typescript/src/game-logic/Food.ts
import { Coordinates } from "./types";
import { Snake } from "./Snake"; // Import Snake to check its body for food placement

export class Food {
  position: Coordinates;

  constructor(boardWidth: number, boardHeight: number, snakeBody: Coordinates[]) {
    this.position = this.generateNewPosition(boardWidth, boardHeight, snakeBody);
  }

  generateNewPosition(boardWidth: number, boardHeight: number, snakeBody: Coordinates[]): Coordinates {
    let newPosition: Coordinates;
    let collisionWithSnake: boolean;

    do {
      newPosition = {
        x: Math.floor(Math.random() * boardWidth),
        y: Math.floor(Math.random() * boardHeight),
      };

      collisionWithSnake = snakeBody.some(
        (segment) => segment.x === newPosition.x && segment.y === newPosition.y
      );
    } while (collisionWithSnake);

    this.position = newPosition;
    return newPosition;
  }
}
